Adding boilerplate about accessing data on S3 (#12)
hubverse-org/hub-dashboard-template: pages/data.qmd

---

> title: "Data"

Reviewer: ```suggestion
title: "Accessing Hub Data"
```

hub-bucket-name: "[hub-bucket-name]"
---

## Accessing hub data on the cloud

> *\[Remove this section if you're not planning to host your hub's data on the Hubverse's AWS infrastructure\]*

Reviewer: ```suggestion
<!-- Remove this page if you're not planning to host your hub's data on the Hubverse's AWS infrastructure -->
```

> To ensure greater access to the data created by and submitted to this hub, real-time copies of its model-output, target, and configuration files are hosted on the Hubverse's Amazon Web Services (AWS) infrastructure, in a public S3 bucket: *\[insert S3 bucket name, as specified in admin.json cloud.host.storage_location\]*

Reviewer: ```suggestion
To ensure greater access to the data created by and submitted to this hub, real-time copies of its model-output, target, and configuration files are hosted on the Hubverse's Amazon Web Services (AWS) infrastructure, in a public S3 bucket: `{{< meta hub-bucket-name >}}`
```

**Note**: For efficient storage, all model-output files in S3 are stored in parquet format, even if the original versions in the GitHub repository are .csv.

GitHub remains the primary interface for operating the hub and collecting forecasts from modelers. However, the mirrors of hub files on S3 are the most convenient way to access hub data without using git/GitHub or cloning the entire hub to your local machine.

The sections below provide examples for accessing hub data on the cloud, depending on your goals and preferred tools. The options include:

| Access Method | Description |
|----|----|
| hubData (R) | Hubverse R client and R code for accessing hub data |
| Polars (Python) | Python open-source library for data manipulation |
| AWS command line interface | Download hub data to your machine and use hubData or Polars for local access |

In general, accessing the data directly from S3 (instead of downloading it first) is more convenient. However, if performance is critical (for example, you're building an interactive visualization), or if you need to work offline, we recommend downloading the data first.

> 
> <!-------------------------------------------------- hubData ------------------------------------------------------->
> 
> <details>
> 
> <summary>hubData (R)</summary>

Reviewer: Since we are rendering quarto here, we can use [a panel tabset](https://quarto.org/docs/interactive/layout.html#tabset-panel). 

```suggestion

:::{#methods .panel-tabset}
<!-------------------------------------------------- hubData ------------------------------------------------------->

## hubData (R)
```

### Using hubData

hubData's [`connect_hub()` function](https://hubverse-org.github.io/hubData/reference/connect_hub.html) returns an [Arrow multi-file dataset](https://arrow.apache.org/docs/r/reference/Dataset.html) that represents a hub's model output data. The dataset can be filtered and transformed using dplyr and then materialized into a local data frame using the [`collect_hub()` function](https://hubverse-org.github.io/hubData/reference/collect_hub.html).

#### Accessing target data

*\[hubData will be updated to access target data once the Hubverse target data standards are finalized.\]*

#### Accessing model output data

Below is an example of using hubData to connect to a hub on S3 and filter the model output data.

``` r
library(dplyr)
library(hubData)

bucket_name <- "hub-bucket-name"
hub_bucket <- s3_bucket(bucket_name)
hub_con <- hubData::connect_hub(hub_bucket, file_format = "parquet", skip_checks = TRUE)
hub_con %>%
  dplyr::filter(location == "MA", output_type == "quantile") %>%
  hubData::collect_hub()
```

-   [full hubData documentation](https://hubverse-org.github.io/hubData/)

> </details>
> 
> <!--------------------------------------------------- Polars ------------------------------------------------------->
> 
> <details>
> 
> <summary>Polars (Python)</summary>

Reviewer: ```suggestion

<!--------------------------------------------------- Polars ------------------------------------------------------->

## Polars (Python)
```

)

# filter and transform as needed and collect into a dataframe, for example:
oracle_dataframe = oracle_data.filter(pl.col("location") == "MA").collect()
```

#### Accessing model output data

Get the model-output files for a specific team (all rounds). This example uses [glob patterns to read from data multiple files into a single dataset](https://docs.pola.rs/user-guide/io/multiple/#reading-into-a-single-dataframe).

``` python
import polars as pl

lf = pl.scan_parquet(
    "s3://{{< meta hub-bucket-name >}}/model-output/[modeling team name]/*.parquet",
    storage_options={"skip_signature": "true"}
)
```

#### Using partitions (hive-style)

If your data uses hive-style partitioning, Polars can use the partitions to filter the data before reading it.

``` python
from datetime import datetime
import polars as pl

oracle_data = pl.scan_parquet(
    "s3://{{< meta hub-bucket-name >}}/target-data/oracle-output/",
    hive_partitioning=True,
    storage_options={"skip_signature": "true"}) \
.filter(pl.col("nowcast_date") == datetime(2025, 2, 5)) \
.collect()
```

-   [Full documentation of the Polars Python API](https://docs.pola.rs/api/python/stable/reference/)

> </details>
> 
> <!--------------------------------------------------- AWS CLI ------------------------------------------------------->
> 
> <details>
> 
> <summary>AWS CLI</summary>

Reviewer: ```suggestion
<!--------------------------------------------------- AWS CLI ------------------------------------------------------->

## AWS CLI
```

AWS provides a terminal-based command line interface (CLI) for exploring and downloading S3 files. This option is ideal if you:

-   plan to work with hub data offline but don't want to use git or GitHub
-   want to download a subset of the data (instead of the entire hub)
-   are using the data for an application that requires local storage or fast response times

### Installing the AWS CLI

-   Install the AWS CLI using the [instructions here](https://docs.aws.amazon.com/cli/latest/userguide/getting-started-install.html)
-   You can skip the instructions for setting up security credentials, since Hubverse data is public

### Using the AWS CLI

When using the AWS CLI, the `--no-sign-request` option is required, since it tells AWS to bypass a credential check (*i.e.*, `--no-sign-request` allows anonymous access to public S3 data).

> > \[!NOTE\] Files in the bucket's `raw` directory should not be used for analysis (they're for internal use only).

Reviewer: ```suggestion
:::{.callout-note}

Files in the bucket's `raw` directory should not be used for analysis (they're for internal use only).

:::
```

List all directories in the hub's S3 bucket:

``` sh
aws s3 ls {{< meta hub-bucket-name >}} --no-sign-request
```

List all files in the hub's bucket:

``` sh
aws s3 ls {{< meta hub-bucket-name >}} --recursive --no-sign-request
```

Download all of target-data contents to your current working directory:

``` sh
aws s3 cp s3://{{< meta hub-bucket-name >}}/target-data/ . --recursive --no-sign-request
```

Download the model-output files for a specific team:

``` sh
aws s3 cp s3://{{< meta hub-bucket-name >}}/[modeling-team-name]/UMass-flusion/ . --recursive --no-sign-request
```

-   [Full documentation for `aws s3 ls`](https://docs.aws.amazon.com/cli/latest/reference/s3/ls.html)
-   [Full documentation for `aws s3 cp`](https://docs.aws.amazon.com/cli/latest/reference/s3/cp.html)

> </details>

Reviewer: closing the panel tabset

```suggestion
:::
```
